### PR TITLE
furtherhance: init at 1.6.0

### PR DESCRIPTION
**Pull request by CaptainJawZ**
###### Description of changes

Hey guys, this is my first contribution, I hope I'm doing it right, this is for the app https://github.com/lakoliu/Furtherance

> Furtherance is a time tracking app written in Rust with GTK 4. It allows you to track time spent on different activities without worrying about your data being captured and sold.

The package as I'm submitting it, builds, installs, and works perfectly! So I hope is ready for others to test it. 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by CaptainJawZ** [comment]
I'm completely lost about the EditorConfig test failing, if anyone could help me pointing me on the right direction I'll fix it ASAP.

**Review by winterqt — CHANGES_REQUESTED** [review]
Please revert the formatting changes you did to both the maintainer list and the package list.

While you're at it, please make these changes into two commits -- first adding you to the maintainer list, and then adding the package (please follow the [commit naming guide](https://nixos.org/manual/nixpkgs/stable/#submitting-changes-making-patches)).

**Comment by CaptainJawZ** [comment]
I think it's done, sorry if I made a mess of git commits, I'm a hyper-mega n00b with git.

**Comment by winterqt** [comment]
You removed a lot of packages and maintainers. How did you rebase?

**Comment by CaptainJawZ** [comment]
Ok what's with this new issue? it's complaining about package outputs that i did not modify.

**Comment by CaptainJawZ** [comment]
I did a git reset/restore on the original fork i did for the files mantainer-list and all-packages then did a rebase, I think that made those files be sent way back into the past.

**Comment by winterqt** [comment]
See the changes tab -- you removed entires from the maintainers list and the packages list.

**Comment by winterqt** [comment]
Okay, cool, just add yourself as a maintainer and the package back to the respective files and we can go from there.

**Comment by CaptainJawZ** [comment]
I think it's all in order now, I hope I'm very sorry and ashamed that i turned this into a commits nightmare. This is my first time.

**Comment by winterqt** [comment]
You managed to add e30273467ebc97415ed248b63d1284ab8bc23e4c somehow. Was this an artifact of the rebase or something?

I'm not sure if trying to salvage this branch is really worthwhile... at some point I'll suggest that you [add an upstream remote](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/configuring-a-remote-for-a-fork) and `git reset --hard upstream/master`, then add your commits and force push.

I have write permissions to your fork, if you'd like me to try and fix this without having to screw around with git yourself. Feel free to try to fix this yourself, though, your work is still here, so you won't lose anything if things don't work out.

**Comment by CaptainJawZ** [comment]
I think I did the hard reset correctly, somehow it closed while I did so, but I think this is all on a clean state and the changes are on the 3 commits that should've been from the start.

**Comment by CaptainJawZ** [comment]
Done as you suggested! did another clean reset then added the commits with those messages, and the appropiate files. Thank you so much for your patience and I'm so sorry for all the trouble.

File: pkgs/applications/misc/furtherance/default.nix
Comment (winterqt) [inline pkgs/applications/misc/furtherance/default.nix:5]: We use lowercase `pname`s.

```suggestion
  pname = "furtherance";
```

File: pkgs/applications/misc/furtherance/default.nix
Comment (winterqt) [inline pkgs/applications/misc/furtherance/default.nix:10]: If the canonical repository name has an uppercase in it, though, we should probably use that.

```suggestion
    repo = "Furtherance";
```

File: pkgs/applications/misc/furtherance/default.nix
Comment (winterqt) [inline pkgs/applications/misc/furtherance/default.nix:36]: Remove this line if it can't be enabled.

**Review by winterqt — CHANGES_REQUESTED** [review]

**Comment by winterqt** [comment]
I messed up the grouping of these suggestions, sorry about that.

File: pkgs/applications/misc/furtherance/default.nix
Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix:36]: Do you mean the # doCheck comment? or the buildInputs? the doCheck can be removed, I was just toying around with that setting.

Comment (winterqt) [reply pkgs/applications/misc/furtherance/default.nix:36]: Yes, `doCheck`. Looking at the application, it doesn't seem to have any tests, so you can remove it just fine (unless I'm wrong, please let me know if that's the case).

Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix]: Yeah that line isnt necessary! I should've removed it from the start, sorry about that! pushed the commit, I am incredibly thankful for your patience helping me with all of this! pushed the commit with all your requested changes.

**Comment by winterqt** [comment]
Please squash all of these commits into 2 commits, [as I described earlier](https://github.com/NixOS/nixpkgs/pull/182405#issuecomment-1192143961). You can use an interactive rebase, or reset the branch + commit.

**Comment by CaptainJawZ** [comment]
Ready!

**Comment by winterqt** [comment]
There's two more small changes that would be beneficial. I'm going to try to walk you through doing a rebase, so you can get an idea for what they look like.

1. Add a newline between lines 3 and 4 of the package, so there's a space between the attrset and the `mkDerivation` call. (This is standard formatting if you look around)
2. Commit this change, the message doesn't matter
3. Run `git rebase -i 9a56cdb0c9519dee20f0c9e7572efd45df7c631e`
4. In the editor that opens, you should see three `pick` lines associated with your two commits. Swap the order of the first and the second, so the maintainer addition is first, and the package addition is second
5. On the third line, for your formatting fixup, change `pick` to `fixup`, so the order is pick `maintainers: ...`, pick `furtherance: init at ...`, and then fixup `...`
6. Save + close the file, and Git should say that the rebase was successful
7. Force push

**Comment by CaptainJawZ** [comment]
That was a bit intimidating at first, but easier to do than I expected! I hope it worked as intended, so if hypotetically, if I made all the commits from last night a "fixup" none of this whould've happened? that's handy. Thank you. Before contributing my next pkgs for PureRef I'll definitely familiarize myself as much as I can with git so this doesn't happen again.

File: pkgs/applications/misc/furtherance/default.nix
Comment (winterqt) [inline pkgs/applications/misc/furtherance/default.nix:19]: Looks like this is incorrect, according to the build logs. (Any clue what happened on your end to cause this, out of curiosity?)

**Review by winterqt — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/furtherance/default.nix
Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix]: That is very odd, other than changing the capitalization of pname, I can't think of any significant changes, but this is coming from cargo, I adjusted the hash and now it builds again. 

I did the rebase again after the fixup, did i do it correctly?

Comment (winterqt) [reply pkgs/applications/misc/furtherance/default.nix]: > other than changing the capitalization of pname

I tested this, and that was the issue.

> I did the rebase again after the fixup, did i do it correctly?

Yup!

File: pkgs/top-level/all-packages.nix
Comment (winterqt) [inline pkgs/top-level/all-packages.nix:26672]: Use one space between attrset braces.

```suggestion
  furtherance = callPackage ../applications/misc/furtherance { };
```

**Review by winterqt — CHANGES_REQUESTED** [review]
Last change, I promise (sorry!)

File: pkgs/top-level/all-packages.nix
Comment (CaptainJawZ) [reply pkgs/top-level/all-packages.nix:26672]: Donion rings! 
I'm sorry for that typo, didn't even noticed when I typed it.

**Review by winterqt — APPROVED** [review]

**Comment by winterqt** [comment]
Just gonna wait on CI to pass, thanks for your patience (and hopefully you learned some Git along the way!).

Out of curiosity, did you do anything different while rebasing? If you squash a commit into another via the `fixup` command, that usually retains the timestamp info on the previous commit, but it seems that your package addition commit got a new timestamp. (It doesn't matter, I'm just curious for whatever reason.)

**Comment by CaptainJawZ** [comment]
I accidentally pushed the fix commit, instead of rebasing first. Thank you so much so so so much for teaching me all of this! I promise my next contribution to be perfect from the beginning!

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/furtherance/default.nix:35]: Might be nice to have this one dependency per line as well so that `git blame` in the future works well.

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/furtherance/default.nix:37]: Meson does this by default.

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/furtherance/default.nix:39]: Conventionally, these should be in `postPatch` so that any checks Meson does see existing shebangs.

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/furtherance/default.nix:43]: This substitution is not necessary, it is only used as a [fallback](https://github.com/lakoliu/Furtherance/blob/996f669796e83f08fd92506441e0340c0f393f61/build-aux/meson/postinstall.py#L6) – Meson [will](https://mesonbuild.com/Reference-manual_builtin_meson.html#mesonadd_install_script) [always](https://github.com/mesonbuild/meson/blob/2e3ac3eec00f32dabcfc470cd4c1090303d08e58/mesonbuild/minstall.py#L646) set `MESON_INSTALL_PREFIX` correctly.

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/furtherance/default.nix:45]: If you are up to it, it would be nice to switch to Meson’s new post install mechanism upstream so this whole thing can be dropped on next release. Example: https://gitlab.gnome.org/GNOME/d-spy/-/merge_requests/6/diffs

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/furtherance/default.nix:51]: `gpl3` is unclear, use `gpl3Plus` or `gpl3Only`. See https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206 for more details.

File: pkgs/applications/misc/furtherance/default.nix
Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix:39]: I'm applying the suggestions! I'm just confused about this and another one I'll comment in a bit. 

So you're suggesting in here that  i replace preInstall with postPatch = ''?

File: pkgs/applications/misc/furtherance/default.nix
Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix:45]: This is something that would need to be modified on the source code of furtherance right? I'm not the developer, just packaging it for nix!

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [reply pkgs/applications/misc/furtherance/default.nix:39]: Yes.

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [reply pkgs/applications/misc/furtherance/default.nix:45]: Right. Often, upstream developers are not aware of the constantly evolving build system best practices and this is where people packaging can help. Making changes upstream can lead to shorter and cleaner packages for Nixpkgs and other distributions.

Comment (winterqt) [reply pkgs/applications/misc/furtherance/default.nix:45]: I've made a [patch for this](https://gist.github.com/winterqt/a596b439016e65b7cceee3a299392128), and I've confirmed it builds. @CaptainJawZ do you mind testing that this runs as expected? You can test this by adding it to `patches` (using `fetchpatch` or a local path) and deleting that entire `postinstall` patch block. If everything works as expected, I'll submit the patch upstream.

Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix:49]: Thank you so much for the help! I'll try it out right now, I made a rebase with all the suggested changes except this one because postInstall doesn't let the app to build, I'll test out this patch.

File: pkgs/applications/misc/furtherance/default.nix
Comment (winterqt) [inline pkgs/applications/misc/furtherance/default.nix:17]: Please move/rename the patch:
```suggestion
    ./gnome-post-install.patch
```

**Review by winterqt — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/furtherance/default.nix
Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix]: Alright, the patch works perfectly thank you so very much for the contribution! this also helps me prevent what would've been a problem with the other package I'm working on, now I know how to properly patch!

with this, I think all the suggestions have been implemented.

I have one question tho, version 1.5.3 is out now, should I wait till this current init is merged into the repo, or adjust it so this builds 1.5.3 instead of 1.5.1? 

It doesn't have any major changes other than a new icon and some translations to french, so in my opinion it can wait.

File: pkgs/applications/misc/furtherance/default.nix
Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix]: done!

File: pkgs/applications/misc/furtherance/default.nix
Comment (winterqt) [reply pkgs/applications/misc/furtherance/default.nix]: > or adjust it so this builds 1.5.3 instead of 1.5.1?

This would be preferable IMO, considering all you should have to do is change version + hashes + commit message.

Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix]: working on it then!

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/furtherance/default.nix:17]:
```suggestion
    # https://github.com/lakoliu/Furtherance/issues/79
    ./gnome-post-install.patch
```

Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix:18]: So in future builds, when this change is implemented on furtherance I just get rid of these lines, right?

**Comment by CaptainJawZ** [comment]
Updated to the newest release!

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/furtherance/default.nix:28]: Is `python3` still necessary?

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [inline pkgs/applications/misc/furtherance/default.nix:46]: This does not seem to be the case: https://github.com/lakoliu/Furtherance/blob/ec40d1db829b829a42e0d45af499b925b59fc983/data/com.lakoliu.Furtherance.appdata.xml.in#L11

Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix:46]: So it would just be licenses.gpl3; ?

File: pkgs/applications/misc/furtherance/default.nix
Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix]: You're right, built it without it and works without issues.

File: pkgs/applications/misc/furtherance/default.nix
Comment (jtojnar) [reply pkgs/applications/misc/furtherance/default.nix]: Right, `gpl3Plus`.

Comment (CaptainJawZ) [reply pkgs/applications/misc/furtherance/default.nix]: Your requests should be present on the latest commit!

**Comment by BenediktBroich** [comment]
Result of `nixpkgs-review pr 182405` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>furtherance</li>
  </ul>
</details>

**Review by BenediktBroich — APPROVED** [review]

**Comment by jtojnar** [comment]
Thanks.

File: pkgs/applications/misc/furtherance/default.nix
Comment (iblech) [reply pkgs/applications/misc/furtherance/default.nix]: Exactly, yes. :-)

**Comment by iblech** [comment]
Sorry for the noise, just want to drop by to express thanks:

Thank you @winterqt, @jtojnar and @BenediktBroich, for patiently welcoming our new contributor to our community and giving this exhaustive review! Thank you, @CaptainJawZ, for contributing this pull request and diligently working on the suggestions!

**Comment by CaptainJawZ** [comment]
@iblech This was an amazing experience for me, and I look forward to my next contributions! Thank you so much for everyone specially @winterqt for teaching me how to properly use Git :D, I will forever be grateful.

**Pull request merged**
